## Parte introductoria
- Definicion de librerias
- Obtencion de datos
- Analisis de datos presentes

In [1]:
import os
import numpy as np 
import pandas as pd 
import matplotlib.pyplot as plt
import matplotlib 

In [2]:
import os

directorio_actual = os.getcwd()
print("Directorio actual:", directorio_actual)


Directorio actual: C:\Users\Anton\Desktop\TFM


In [3]:
df=pd.read_csv("DatosCancer/heart.csv")
df.head()

Age Sex ChestPainType  RestingBP  Cholesterol  FastingBS RestingECG  MaxHR  \
0   40   M           ATA        140          289          0     Normal    172   
1   49   F           NAP        160          180          0     Normal    156   
2   37   M           ATA        130          283          0         ST     98   
3   48   F           ASY        138          214          0     Normal    108   
4   54   M           NAP        150          195          0     Normal    122   

  ExerciseAngina  Oldpeak ST_Slope  HeartDisease  
0              N      0.0       Up             0  
1              N      1.0     Flat             1  
2              N      0.0       Up             0  
3              Y      1.5     Flat             1  
4              N      0.0       Up             0

In [4]:
df.dtypes

Age                 int64
Sex                object
ChestPainType      object
RestingBP           int64
Cholesterol         int64
FastingBS           int64
RestingECG         object
MaxHR               int64
ExerciseAngina     object
Oldpeak           float64
ST_Slope           object
HeartDisease        int64
dtype: object

In [5]:
string_col = df.select_dtypes(include="object").columns
df[string_col]=df[string_col].astype("string")
df.dtypes

Age                 int64
Sex                string
ChestPainType      string
RestingBP           int64
Cholesterol         int64
FastingBS           int64
RestingECG         string
MaxHR               int64
ExerciseAngina     string
Oldpeak           float64
ST_Slope           string
HeartDisease        int64
dtype: object

In [6]:
df.describe().T

count        mean         std   min     25%    50%    75%    max
Age           918.0   53.510893    9.432617  28.0   47.00   54.0   60.0   77.0
RestingBP     918.0  132.396514   18.514154   0.0  120.00  130.0  140.0  200.0
Cholesterol   918.0  198.799564  109.384145   0.0  173.25  223.0  267.0  603.0
FastingBS     918.0    0.233115    0.423046   0.0    0.00    0.0    0.0    1.0
MaxHR         918.0  136.809368   25.460334  60.0  120.00  138.0  156.0  202.0
Oldpeak       918.0    0.887364    1.066570  -2.6    0.00    0.6    1.5    6.2
HeartDisease  918.0    0.553377    0.497414   0.0    0.00    1.0    1.0    1.0

Age: age of the patient [years]

Sex: sex of the patient [M: Male, F: Female]

ChestPainType: chest pain type [TA: Typical Angina, ATA: Atypical Angina, NAP: Non-Anginal Pain, ASY: Asymptomatic]

RestingBP: resting blood pressure [mm Hg]

Cholesterol: serum cholesterol [mm/dl]

FastingBS: fasting blood sugar [1: if FastingBS > 120 mg/dl, 0: otherwise]

RestingECG: resting electrocardiogram results [Normal: Normal, ST: having ST-T wave abnormality (T wave inversions and/or ST elevation or depression of > 0.05 mV), LVH: showing probable or definite left ventricular hypertrophy by Estes' criteria]

MaxHR: maximum heart rate achieved [Numeric value between 60 and 202]

ExerciseAngina: exercise-induced angina [Y: Yes, N: No]

Oldpeak: oldpeak = ST [Numeric value measured in depression]

ST_Slope: the slope of the peak exercise ST segment [Up: upsloping, Flat: flat, Down: downsloping]

HeartDisease: output class [1: heart disease, 0: Normal]

## Preprocesamiento de datos


In [7]:
# Buscar si hay valores NULL
df.isnull().sum()

Age               0
Sex               0
ChestPainType     0
RestingBP         0
Cholesterol       0
FastingBS         0
RestingECG        0
MaxHR             0
ExerciseAngina    0
Oldpeak           0
ST_Slope          0
HeartDisease      0
dtype: int64

In [8]:
# Tratamiento de variables categoricas

df[string_col].head()
for col in string_col:
    print(f"La distribución de valores categóricos en la columna {col} es : ")
    print(df[col].value_counts())


La distribución de valores categóricos en la columna Sex es : 
M    725
F    193
Name: Sex, dtype: Int64
La distribución de valores categóricos en la columna ChestPainType es : 
ASY    496
NAP    203
ATA    173
TA      46
Name: ChestPainType, dtype: Int64
La distribución de valores categóricos en la columna RestingECG es : 
Normal    552
LVH       188
ST        178
Name: RestingECG, dtype: Int64
La distribución de valores categóricos en la columna ExerciseAngina es : 
N    547
Y    371
Name: ExerciseAngina, dtype: Int64
La distribución de valores categóricos en la columna ST_Slope es : 
Flat    460
Up      395
Down     63
Name: ST_Slope, dtype: Int64


In [9]:
df_numerical=pd.get_dummies(df,columns=string_col,drop_first=False)
df_numerical.head()

Age  RestingBP  Cholesterol  FastingBS  MaxHR  Oldpeak  HeartDisease  \
0   40        140          289          0    172      0.0             0   
1   49        160          180          0    156      1.0             1   
2   37        130          283          0     98      0.0             0   
3   48        138          214          0    108      1.5             1   
4   54        150          195          0    122      0.0             0   

   Sex_F  Sex_M  ChestPainType_ASY  ...  ChestPainType_NAP  ChestPainType_TA  \
0      0      1                  0  ...                  0                 0   
1      1      0                  0  ...                  1                 0   
2      0      1                  0  ...                  0                 0   
3      1      0                  1  ...                  0                 0   
4      0      1                  0  ...                  1                 0   

   RestingECG_LVH  RestingECG_Normal  RestingECG_ST  ExerciseAngina_N  \
0               0                  1              0                 1   
1               0                  1              0                 1   
2               0                  0              1                 1   
3               0                  1              0                 0   
4               0                  1              0                 1   

   ExerciseAngina_Y  ST_Slope_Down  ST_Slope_Flat  ST_Slope_Up  
0                 0              0              0            1  
1                 0              0              1            0  
2                 0              0              0            1  
3                 1              0              1            0  
4                 0              0              0            1  

[5 rows x 21 columns]

In [10]:
# Normalizacion para eliminar dimension y evitar los sesgos con el uso de unidades.
# En este caso se utilizará MINMAXSCALER
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
data = pd.DataFrame(scaler.fit_transform(df_numerical), columns=df_numerical.columns)
print(data)

          Age  RestingBP  Cholesterol  FastingBS     MaxHR   Oldpeak  \
0    0.244898       0.70     0.479270        0.0  0.788732  0.295455   
1    0.428571       0.80     0.298507        0.0  0.676056  0.409091   
2    0.183673       0.65     0.469320        0.0  0.267606  0.295455   
3    0.408163       0.69     0.354892        0.0  0.338028  0.465909   
4    0.530612       0.75     0.323383        0.0  0.436620  0.295455   
..        ...        ...          ...        ...       ...       ...   
913  0.346939       0.55     0.437811        0.0  0.507042  0.431818   
914  0.816327       0.72     0.320066        1.0  0.570423  0.681818   
915  0.591837       0.65     0.217247        0.0  0.387324  0.431818   
916  0.591837       0.65     0.391376        0.0  0.802817  0.295455   
917  0.204082       0.69     0.290216        0.0  0.795775  0.295455   

     HeartDisease  Sex_F  Sex_M  ChestPainType_ASY  ...  ChestPainType_NAP  \
0             0.0    0.0    1.0                0.0  ...  

## Definicion de Modelo - Regresion logística
Se decide tomar la técnica mencionada ya que permite trabajar con conjuntos de datos en los que la variable dependiente solamente puede tomar un par de valores, lo que generalmente se etiqueta como positivos (P) y negativos (N).

In [11]:
#MODELO BASICO
from sklearn.linear_model import LogisticRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, precision_score, recall_score, confusion_matrix

# Variables predictoras (X): todas las columnas excepto 'HeartDisease'
x_basic = data.drop('HeartDisease', axis=1)

# Variable objetivo (Y): 'HeartDisease'
y_basic = data['HeartDisease']

# Division
x_train_basic, x_test_basic, y_train_basic, y_test_basic = train_test_split(x_basic, y_basic, random_state = 1)

# Ajuste del modelo
classifier_basic = LogisticRegression().fit(x_train_basic, y_train_basic)
print(classifier_basic.get_params())
y_train_pred_basic = classifier_basic.predict(x_train_basic)
y_test_pred_basic = classifier_basic.predict(x_test_basic)

# Mostrar resultados de rendimiento
print('Resultados del conjunto de entrenamiento')
print('Precision: ', accuracy_score(y_train_basic, y_train_pred_basic))
print('Exactitud: ', precision_score(y_train_basic, y_train_pred_basic))
print('Exhaustividad: ', recall_score(y_train_basic, y_train_pred_basic))

print('Resultados del conjunto de test')
print('Precision: ', accuracy_score(y_test_basic, y_test_pred_basic))
print('Exactitud: ', precision_score(y_test_basic, y_test_pred_basic))
print('Exhaustividad: ', recall_score(y_test_basic, y_test_pred_basic))



{'C': 1.0, 'class_weight': None, 'dual': False, 'fit_intercept': True, 'intercept_scaling': 1, 'l1_ratio': None, 'max_iter': 100, 'multi_class': 'auto', 'n_jobs': None, 'penalty': 'l2', 'random_state': None, 'solver': 'lbfgs', 'tol': 0.0001, 'verbose': 0, 'warm_start': False}
Resultados del conjunto de entrenamiento
Precision:  0.8633720930232558
Exactitud:  0.8601583113456465
Exhaustividad:  0.888283378746594
Resultados del conjunto de test
Precision:  0.8826086956521739
Exactitud:  0.9191176470588235
Exhaustividad:  0.8865248226950354


In [12]:
# MODELO CON ANALISIS DE HIPERPARAMETROS
from sklearn.model_selection import GridSearchCV

# Definir el modelo de regresión logística
logistic_classifier = LogisticRegression(max_iter=10000)

# Valores de hiperparámetros a probar
param_grid_logistic = {
    'penalty': ['l1', 'l2', 'elasticnet'],
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'solver': ['newton-cg', 'lbfgs', 'liblinear', 'sag', 'saga']
}

# Configurar la búsqueda con validación cruzada
grid_search_logistic = GridSearchCV(logistic_classifier, param_grid_logistic, cv=5, scoring='accuracy')

# Ajustar el modelo a los datos de entrenamiento
grid_search_logistic.fit(x_train_basic, y_train_basic)

# Obtener el mejor conjunto de hiperparámetros
best_params_logistic = grid_search_logistic.best_params_
print(f"Mejor conjunto de hiperparámetros para Regresión Logística: {best_params_logistic}")

# Utilizar el mejor modelo para predecir en los conjuntos de entrenamiento y prueba
y_train_pred_logistic = grid_search_logistic.predict(x_train_basic)
y_test_pred_logistic = grid_search_logistic.predict(x_test_basic)

# Mostrar resultados de rendimiento con el mejor modelo
print('Resultados del conjunto de entrenamiento con ajuste de hiperparámetros (Regresión Logística)')
print('Precisión: ', accuracy_score(y_train_basic, y_train_pred_logistic))
print('Exactitud: ', precision_score(y_train_basic, y_train_pred_logistic))
print('Exhaustividad: ', recall_score(y_train_basic, y_train_pred_logistic))

print('Resultados del conjunto de prueba con ajuste de hiperparámetros (Regresión Logística)')
print('Precisión: ', accuracy_score(y_test_basic, y_test_pred_logistic))
print('Exactitud: ', precision_score(y_test_basic, y_test_pred_logistic))
print('Exhaustividad: ', recall_score(y_test_basic, y_test_pred_logistic))


Mejor conjunto de hiperparámetros para Regresión Logística: {'C': 0.1, 'penalty': 'l2', 'solver': 'liblinear'}
Resultados del conjunto de entrenamiento con ajuste de hiperparámetros (Regresión Logística)
Precisión:  0.8633720930232558
Exactitud:  0.8545454545454545
Exhaustividad:  0.896457765667575
Resultados del conjunto de prueba con ajuste de hiperparámetros (Regresión Logística)
Precisión:  0.8913043478260869
Exactitud:  0.9202898550724637
Exhaustividad:  0.900709219858156


C:\Users\Anton\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py:372: FitFailedWarning: 
240 fits failed out of a total of 450.
The score on these train-test partitions for these parameters will be set to nan.
If these failures are not expected, you can try to debug them by setting error_score='raise'.

Below are more details about the failures:
--------------------------------------------------------------------------------
30 fits failed with the following error:
Traceback (most recent call last):
  File "C:\Users\Anton\anaconda3\lib\site-packages\sklearn\model_selection\_validation.py", line 680, in _fit_and_score
    estimator.fit(X_train, y_train, **fit_params)
  File "C:\Users\Anton\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 1461, in fit
    solver = _check_solver(self.solver, self.penalty, self.dual)
  File "C:\Users\Anton\anaconda3\lib\site-packages\sklearn\linear_model\_logistic.py", line 447, in _check_solver
    raise ValueError(
Va

In [13]:
# Obtener resultados de la búsqueda de hiperparámetros
results = pd.DataFrame(grid_search_logistic.cv_results_)

# Seleccionar columnas relevantes
columns_of_interest = ['params', 'mean_test_score', 'std_test_score']
results_subset = results[columns_of_interest]

# Mostrar la tabla comparativa
print(results_subset)

                                               params  mean_test_score  \
0   {'C': 0.001, 'penalty': 'l1', 'solver': 'newto...              NaN   
1    {'C': 0.001, 'penalty': 'l1', 'solver': 'lbfgs'}              NaN   
2   {'C': 0.001, 'penalty': 'l1', 'solver': 'libli...         0.466571   
3      {'C': 0.001, 'penalty': 'l1', 'solver': 'sag'}              NaN   
4     {'C': 0.001, 'penalty': 'l1', 'solver': 'saga'}         0.521834   
..                                                ...              ...   
85  {'C': 100, 'penalty': 'elasticnet', 'solver': ...              NaN   
86  {'C': 100, 'penalty': 'elasticnet', 'solver': ...              NaN   
87  {'C': 100, 'penalty': 'elasticnet', 'solver': ...              NaN   
88  {'C': 100, 'penalty': 'elasticnet', 'solver': ...              NaN   
89  {'C': 100, 'penalty': 'elasticnet', 'solver': ...              NaN   

    std_test_score  
0              NaN  
1              NaN  
2         0.002688  
3              NaN  
4     

In [14]:
# datos = {
#             'Age': 49,
#             'Sex': "F",
#             'ChestPainType': "NAP",
#             'RestingBP': 160,
#             'Cholesterol': 180,
#             'FastingBS': 0,
#             'RestingECG': "Normal",
#             'MaxHR': 156,
#             'ExerciseAngina': "N",
#             'Oldpeak': 1,
#             'ST_Slope': "Flat"
#         }

# data_input = pd.DataFrame(datos, index=[0])
# # Añadir la nueva fila al principio del DataFrame original
# df_input = pd.concat([data_input, df], ignore_index=True)
# df_input = df_input.drop(columns=['HeartDisease'])
# string_col = df_input.select_dtypes(include="object").columns
# df_input[string_col]=df_input[string_col].astype("string")

# df_numerical_input=pd.get_dummies(df_input,columns=string_col,drop_first=False)

# scaler_input = MinMaxScaler()
# data = pd.DataFrame(scaler_input.fit_transform(df_numerical_input), columns=df_numerical_input.columns)

# # print(data.iloc[0])
# y_output = classifier.predict(data.head(1))
# print(y_output)




## Definición de Modelo - Máquinas de vector soporte

In [15]:
# MODELO BASICO
from sklearn.svm import SVC


svm_classifier_basic = SVC().fit(x_train_basic, y_train_basic)
print(svm_classifier_basic.get_params())

y_train_pred_SVM_basic = svm_classifier_basic.predict(x_train_basic)
y_test_pred_SVM_basic = svm_classifier_basic.predict(x_test_basic)

# Mostrar resultados de rendimiento
print('Resultados del conjunto de entrenamiento')
print('Precisión: ', accuracy_score(y_train_basic,y_train_pred_SVM_basic))
print('Exactitud: ', precision_score(y_train_basic,y_train_pred_SVM_basic))
print('Exhaustividad: ', recall_score(y_train_basic,y_train_pred_SVM_basic))

print('Resultados del conjunto de test')
print('Precisión: ', accuracy_score(y_test_basic,y_test_pred_SVM_basic))
print('Exactitud: ', precision_score(y_test_basic,y_test_pred_SVM_basic))
print('Exhaustividad: ', recall_score(y_test_basic,y_test_pred_SVM_basic))

{'C': 1.0, 'break_ties': False, 'cache_size': 200, 'class_weight': None, 'coef0': 0.0, 'decision_function_shape': 'ovr', 'degree': 3, 'gamma': 'scale', 'kernel': 'rbf', 'max_iter': -1, 'probability': False, 'random_state': None, 'shrinking': True, 'tol': 0.001, 'verbose': False}
Resultados del conjunto de entrenamiento
Precisión:  0.872093023255814
Exactitud:  0.8567774936061381
Exhaustividad:  0.9128065395095368
Resultados del conjunto de test
Precisión:  0.8913043478260869
Exactitud:  0.9084507042253521
Exhaustividad:  0.9148936170212766


In [16]:
# MODELO CON AJUSTE DE HIPERPARÁMETROS
from sklearn.model_selection import GridSearchCV

# Definir el modelo de SVM
svm_classifier = SVC()

# Valores de hiperparámetros a probar
param_grid_svm = {
    'C': [0.001, 0.01, 0.1, 1, 10, 100],
    'kernel': ['linear', 'poly', 'rbf', 'sigmoid'],
    'gamma': ['scale', 'auto', 0.001, 0.01, 0.1, 1, 10, 100]
}

# Configurar la búsqueda con validación cruzada
grid_search_svm = GridSearchCV(svm_classifier, param_grid_svm, cv=5, scoring='accuracy')

# Ajustar el modelo a los datos de entrenamiento
grid_search_svm.fit(x_train_basic, y_train_basic)

# Obtener el mejor conjunto de hiperparámetros
best_params_svm = grid_search_svm.best_params_
print(f"Mejor conjunto de hiperparámetros para SVM: {best_params_svm}")

# Utilizar el mejor modelo para predecir en los conjuntos de entrenamiento y prueba
y_train_pred_svm = grid_search_svm.predict(x_train_basic)
y_test_pred_svm = grid_search_svm.predict(x_test_basic)

# Mostrar resultados de rendimiento con el mejor modelo
print('Resultados del conjunto de entrenamiento con ajuste de hiperparámetros (SVM)')
print('Precisión: ', accuracy_score(y_train_basic, y_train_pred_svm))
print('Exactitud: ', precision_score(y_train_basic, y_train_pred_svm))
print('Exhaustividad: ', recall_score(y_train_basic, y_train_pred_svm))

print('Resultados del conjunto de prueba con ajuste de hiperparámetros (SVM)')
print('Precisión: ', accuracy_score(y_test_basic, y_test_pred_svm))
print('Exactitud: ', precision_score(y_test_basic, y_test_pred_svm))
print('Exhaustividad: ', recall_score(y_test_basic, y_test_pred_svm))


Mejor conjunto de hiperparámetros para SVM: {'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}
Resultados del conjunto de entrenamiento con ajuste de hiperparámetros (SVM)
Precisión:  0.8706395348837209
Exactitud:  0.8696808510638298
Exhaustividad:  0.8910081743869209
Resultados del conjunto de prueba con ajuste de hiperparámetros (SVM)
Precisión:  0.8913043478260869
Exactitud:  0.9142857142857143
Exhaustividad:  0.9078014184397163


## Definición de Modelo - Árboles de decisión

In [17]:
from sklearn.tree import DecisionTreeClassifier
from sklearn.preprocessing import LabelEncoder

# Transformar las columnas categóricas en valores numéricos
df_tree = df.apply(LabelEncoder().fit_transform)
df_tree.head()

scaler_tree = MinMaxScaler()
data_tree = pd.DataFrame(scaler_tree.fit_transform(df_tree), columns=df_tree.columns)

# Variables predictoras (X): todas las columnas excepto 'HeartDisease'
x_tree = data_tree.drop('HeartDisease', axis=1)

# Variable objetivo (Y): 'HeartDisease'
y_tree = data_tree['HeartDisease']

# Division
x_tree_train, x_tree_test, y_tree_train, y_tree_test = train_test_split(x_tree, y_tree, random_state = 1)


DecisionTree_train=DecisionTreeClassifier(criterion="entropy").fit(x_tree_train,y_tree_train)
DecisionTree_test=DecisionTreeClassifier(criterion="entropy").fit(x_tree_test,y_tree_test)
y_train_pred_DecisionTree=DecisionTree_train.predict(x_tree_train)
y_test_pred_DecisionTree = DecisionTree_test.predict(x_tree_test)


# Mostrar resultados de rendimiento
print('Resultados del conjunto de entrenamiento')
print('Precisión: ', accuracy_score(y_tree_train,y_train_pred_DecisionTree))
print('Exactitud: ', precision_score(y_tree_train,y_train_pred_DecisionTree))
print('Exhaustividad: ', recall_score(y_tree_train,y_train_pred_DecisionTree))

print('Resultados del conjunto de test')
print('Precisión: ', accuracy_score(y_tree_test,y_test_pred_DecisionTree))
print('Exactitud: ', precision_score(y_tree_test,y_test_pred_DecisionTree))
print('Exhaustividad: ', recall_score(y_tree_test,y_test_pred_DecisionTree))

Resultados del conjunto de entrenamiento
Precisión:  1.0
Exactitud:  1.0
Exhaustividad:  1.0
Resultados del conjunto de test
Precisión:  1.0
Exactitud:  1.0
Exhaustividad:  1.0


In [18]:
from sklearn.model_selection import GridSearchCV

# Definir el modelo de árbol de decisiones
tree_classifier = DecisionTreeClassifier()

# Valores de hiperparámetros a probar
param_grid = {
    'criterion': ['gini', 'entropy'],
    'max_depth': [None, 1, 2, 3, 4, 5, 6, 7, 8, 10, 12, 14, 16, 18, 20, 22, 24, 26, 29, 30],
    'min_samples_split': [2, 5, 10],
    'min_samples_leaf': [1, 2, 4]
}

# Configurar la búsqueda con validación cruzada
grid_search = GridSearchCV(tree_classifier, param_grid, cv=5, scoring='accuracy')

# Ajustar el modelo a los datos de entrenamiento
grid_search.fit(x_tree_train, y_tree_train)

# Obtener el mejor conjunto de hiperparámetros
best_params = grid_search.best_params_
print(f"Mejor conjunto de hiperparámetros: {best_params}")

# Utilizar el mejor modelo para predecir en los conjuntos de entrenamiento y prueba
y_train_pred_best = grid_search.predict(x_tree_train)
y_test_pred_best = grid_search.predict(x_tree_test)

# Mostrar resultados de rendimiento con el mejor modelo
print('Resultados del conjunto de entrenamiento con ajuste de hiperparámetros')
print('Precisión: ', accuracy_score(y_tree_train, y_train_pred_best))
print('Exactitud: ', precision_score(y_tree_train, y_train_pred_best))
print('Exhaustividad: ', recall_score(y_tree_train, y_train_pred_best))

print('Resultados del conjunto de prueba con ajuste de hiperparámetros')
print('Precisión: ', accuracy_score(y_tree_test, y_test_pred_best))
print('Exactitud: ', precision_score(y_tree_test, y_test_pred_best))
print('Exhaustividad: ', recall_score(y_tree_test, y_test_pred_best))


Mejor conjunto de hiperparámetros: {'criterion': 'entropy', 'max_depth': 5, 'min_samples_leaf': 4, 'min_samples_split': 5}
Resultados del conjunto de entrenamiento con ajuste de hiperparámetros
Precisión:  0.8880813953488372
Exactitud:  0.8897849462365591
Exhaustividad:  0.9019073569482289
Resultados del conjunto de prueba con ajuste de hiperparámetros
Precisión:  0.8521739130434782
Exactitud:  0.8848920863309353
Exhaustividad:  0.8723404255319149


## Conclusión

In [25]:
print('Regresión logística:')
print('Básico - Precisión: ', (accuracy_score(y_train_basic, y_train_pred_basic)+accuracy_score(y_test_basic, y_test_pred_basic))/2, ' Analizado - Precisión: ',(accuracy_score(y_train_basic, y_train_pred_logistic)+accuracy_score(y_test_basic, y_test_pred_logistic))/2)
print('Básico - Exactitud: ', (precision_score(y_train_basic, y_train_pred_basic)+precision_score(y_test_basic, y_test_pred_basic))/2, ' Analizado - Exactitud: ',(precision_score(y_train_basic, y_train_pred_logistic)+precision_score(y_test_basic, y_test_pred_logistic))/2)
print('Básico - Exhaustividad: ', (recall_score(y_train_basic, y_train_pred_basic)+recall_score(y_test_basic, y_test_pred_basic))/2, ' Analizado - Exhaustividad: ',(recall_score(y_train_basic, y_train_pred_logistic)+recall_score(y_test_basic, y_test_pred_logistic))/2)

print('Máquinas de vectores de soporte:')
print('Básico - Precisión: ', (accuracy_score(y_train_basic, y_train_pred_SVM_basic)+accuracy_score(y_test_basic, y_test_pred_SVM_basic))/2, ' Analizado - Precisión: ',(accuracy_score(y_train_basic, y_train_pred_svm)+accuracy_score(y_test_basic, y_test_pred_svm))/2)
print('Básico - Exactitud: ', (precision_score(y_train_basic, y_train_pred_SVM_basic)+precision_score(y_test_basic, y_test_pred_SVM_basic))/2, ' Analizado - Exactitud: ',(precision_score(y_train_basic, y_train_pred_svm)+precision_score(y_test_basic, y_test_pred_svm))/2)
print('Básico - Exhaustividad: ', (recall_score(y_train_basic, y_train_pred_SVM_basic)+recall_score(y_test_basic, y_test_pred_SVM_basic))/2, ' Analizado - Exhaustividad: ',(recall_score(y_train_basic, y_train_pred_svm)+recall_score(y_test_basic, y_test_pred_svm))/2)

print('Árboles de decisión:')
print('Analizado - Precisión: ', (accuracy_score(y_tree_train, y_train_pred_best)+accuracy_score(y_tree_test, y_test_pred_best))/2)
print('Analizado - Exactitud: ', (precision_score(y_tree_train, y_train_pred_best)+precision_score(y_tree_test, y_test_pred_best))/2)
print('Analizado - Exhaustividad: ', (recall_score(y_tree_train, y_train_pred_best)+recall_score(y_tree_test, y_test_pred_best))/2)


Regresión logística:
Básico - Precisión:  0.8729903943377149  Analizado - Precisión:  0.8773382204246714
Básico - Exactitud:  0.889637979202235  Analizado - Exactitud:  0.8874176548089592
Básico - Exhaustividad:  0.8874041007208147  Analizado - Exhaustividad:  0.8985834927628655
Máquinas de vectores de soporte:
Básico - Precisión:  0.8816986855409504  Analizado - Precisión:  0.880971941354904
Básico - Exactitud:  0.8826140989157452  Analizado - Exactitud:  0.891983282674772
Básico - Exhaustividad:  0.9138500782654067  Analizado - Exhaustividad:  0.8994047964133186
Árboles de decisión:
Analizado - Precisión:  0.8701276541961578
Analizado - Exactitud:  0.8873385162837473
Analizado - Exhaustividad:  0.8871238912400718


In [26]:
a = SVC(C=10,gamma=0.01,kernel='rbf')